### PR TITLE
Trigger a warning for prolonged rate limiting

### PR DESCRIPTION
**Pull request by brandon-m-skinner**
## Description

Presently we just push the messages to Slack. We should also trigger a warning so functionality built around that can pick it up and put it where it needs to be.

## Checklist

Please make sure the items below have been covered before requesting a review:

- [x] This change works and has been tested locally (or has an appropriate fallback).
- [x] This change works and has been tested on a Go sandbox.
- [x] This change has relevant unit tests (if applicable).
- [x] This change has relevant documentation additions / updates (if applicable).

## Steps to Test

1. Set the rate limiting thresholds for Search queries and indexing low.
2. Trigger ratelimiting. There should be no warning generated.
3. Apply PR.
4. Trigger ratelimiting. A warning should be generated.

### CODE REVIEW

File: search/includes/classes/class-queue.php [hunk 1]
```diff
@@ -721,13 +721,15 @@ public function maybe_alert_for_prolonged_index_limiting() {
 		}
 
 		$message = sprintf(
-			'Application %d - %s is index limited for %d seconds',
+			'Application %d - %s has had it\'s Elasticsearch indexing rate limited for %d seconds. Large batch indexing operations are being queued for indexing in batches over time.',
```
Comment (nickdaugherty) [inline search/includes/classes/class-queue.php:724]:
```suggestion
			'Application %d - %s has had its Elasticsearch indexing rate limited for %d seconds. Large batch indexing operations are being queued for indexing in batches over time.',
```

File: search/includes/classes/class-search.php [hunk 2]
```diff
@@ -691,13 +691,15 @@ public function maybe_alert_for_prolonged_query_limiting() {
 		}
 
 		$message = sprintf(
-			'Application %d - %s is query limited for %d seconds',
+			'Application %d - %s has had it\'s Elasticsearch queries rate limited for %d seconds. Half of traffic is diverted to the database when queries are rate limited.',
```
Comment (nickdaugherty) [inline search/includes/classes/class-search.php:694]:
```suggestion
			'Application %d - %s has had its Elasticsearch queries rate limited for %d seconds. Half of traffic is diverted to the database when queries are rate limited.',
```

File: tests/search/test-class-search.php [hunk 3]
```diff
@@ -2029,7 +2029,8 @@ public function test__maybe_alert_for_prolonged_query_limiting( $difference, $sh
 			->method( 'send_to_chat' )
 			->with( '#vip-go-es-alerts', $this->anything(), $expected_level );
 
-		$es->maybe_alert_for_prolonged_query_limiting();
+		// Surpress the trigger_error warning since it will always happens and causes execution to fail
+		@$es->maybe_alert_for_prolonged_query_limiting();
```
Comment (nickdaugherty) [inline tests/search/test-class-search.php:2033]: PHPUnit can be told to expect exceptions (which is what eventually happens when the called code triggers an error) - that has the benefit of not using error supression but also asserting that the warning is created:

https://github.com/Automattic/vip-go-mu-plugins/blob/429dc63cffdf82fdbc9f663ce904fb10e09a46b0/tests/search/test-class-search.php#L1267-L1268

**Review by nickdaugherty — CHANGES_REQUESTED** [review]
Looks good, just noted a few small things.

**Review by nickdaugherty — APPROVED** [review]

**Review by netsuso — APPROVED** [review]

**Comment by netsuso** [comment]
stacks-r1280

**Pull request merged**